Implement `newHeads` WebSocket Subscriptions (eth_subscribe)

LogvinovLeon opened the pull request:
# Description

## Key Changes

- **New `Watermark` utility** for tracking monotonically increasing sequences
- **`newHeads` subscription** (new) - streams new block headers
- **`logs` subscription** (refactored) - fixes duplication bug with watermark pattern
- **Parameter validation module** - dedicated `params` module with `SubscriptionRequest` enum
- **Refactored service** - split into logical impl blocks with focused helper functions

-----

- [ ] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [ ] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

Rendered docs are available at `https://sovlabs-ci-rustdoc-artifacts.us-east-1.amazonaws.com/<BRANCH_NAME>/index.html`

LogvinovLeon committed:
Add failing test demonstrating WebSocket .watch() hangs

This test shows that Alloy's .watch() method hangs indefinitely
when using WebSocket client, even with instant finality (finalization=0).
The test times out after 1 second, demonstrating the issue.

This is why the soak testing tool was reverted to use HTTP client instead.

LogvinovLeon committed:
Add test demonstrating newHeads subscription is not implemented

This test shows that WebSocket subscriptions to newHeads do not work,
which is why Alloy's .watch() method hangs. The .watch() method relies
on receiving block notifications via newHeads subscription to detect
when transactions are mined.

Root cause: sov-ethereum only implements 'logs' subscriptions, not
'newHeads' or 'newPendingTransactions' subscriptions.

LogvinovLeon committed:
Fix eth_subscribe parameter parsing to handle standard Ethereum format

Changed from expecting a struct with named fields to accepting positional
parameters matching the standard Ethereum JSON-RPC format:
- eth_subscribe("newHeads")
- eth_subscribe("logs", {filter})

This fixes the parsing error where the server expected 2 elements but
only received 1 when clients tried to subscribe to newHeads.

Note: newHeads subscription is still not implemented (only logs are
supported), so the subscription will be rejected. This commit only
fixes the parameter parsing layer.

LogvinovLeon committed:
Fix eth_subscribe params parsing

LogvinovLeon committed:
Update WebSocket tests to check proper error handling

- ws_watch_returns_receipt: Use .get_receipt() instead of .watch() to verify basic transaction flow works
- ws_subscribe_new_heads: Verify that newHeads subscription returns proper error message

These tests demonstrate that the parameter parsing fix works correctly and subscriptions are properly rejected with clear error messages.

LogvinovLeon committed:
Implement blocks streaming

LogvinovLeon committed:
Update tests of blocks subscription to test success

LogvinovLeon committed:
Better error handling

LogvinovLeon committed:
Refactor WebSocket subscriptions with watermark pattern

Introduce a Watermark utility to track monotonically increasing sequences
and simplify the subscription logic in both logs and blocks handlers.

Key changes:
- Add watermark.rs module with advance() method for tracking high-water marks
- Refactor logs subscription to use watermark for transaction tracking
- Refactor blocks subscription to use watermark for block number tracking
- Extract helper methods get_block() and get_receipt() with better error handling
- Simplify subscription loop logic by eliminating manual index tracking
- Rename send_subscription_message() to send() for brevity

The watermark pattern provides cleaner separation between tracking progression
and handling individual items, making the code more maintainable.

LogvinovLeon committed:
Extract subscription parameter validation into separate module

Move parameter validation logic from subscribe.rs into a dedicated params
module with a cleaner architecture:

- Introduce SubscriptionRequest enum to represent validated subscription types
- Extract validation into params::validate() function with proper error handling
- Simplify subscribe.rs by reducing nested match statements
- Move Error type to params module and implement From<Error> for ErrorObjectOwned
- Reorganize service.rs by moving Error enum to the top of the file
- Reduce code duplication and improve separation of concerns

This refactoring makes the subscription handler more maintainable by clearly
separating validation logic from subscription lifecycle management.

LogvinovLeon committed:
Simplify subscription handler with reject helper

Extract common error handling pattern into a reject() helper function
to reduce duplication and improve readability.

Changes:
- Add parse() function in params module to extract parameter parsing
- Add reject() helper to handle pending.reject() and return pattern
- Simplify error handling in eth_subscribe with early returns
- Remove redundant imports (SubscriptionKind, Params now only in params module)
- Clean up formatting and reduce nesting

The code is now more linear and easier to follow with less boilerplate.

LogvinovLeon committed:
Refactor subscription service for better readability

Reorganize service.rs to improve code organization and readability:

- Split impl block to group public methods separately from helpers
- Place logs() and blocks() entry points adjacent for easy comparison
- Extract send_matching_logs() to reduce nesting in logs()
- Extract send_block_header() to simplify blocks()
- Remove data_type parameter from send() method
- Fix typos in error messages ("eist" -> "exist")
- Add doc comments to public entry points

The two main subscription methods now have a parallel structure that
makes their similarities and differences immediately apparent.

LogvinovLeon committed:
Update CHANGELOG

LogvinovLeon committed:
Use watermark correctly for tx indices as they are exclusive ranges

bkolad commented on crates/full-node/sov-ethereum/src/handlers/subscribe/service.rs:
```diff
@@ -0,0 +1,172 @@
+            let mut state = self.ethereum.api_state_accessor();
+            let pending_block = self.evm.pending_block(&mut state);
+
+            for tx_idx in tx_watermark.advance(pending_block.transactions.end) {
```
It looks like if `pending_block.transactions.end = 7`, we will try to iterate over index 7 here. However, the `ending_block.transactions` range is exclusive, so it appears we’re attempting to access a transaction index that doesn’t exist.

LogvinovLeon replied:
Fixed

bkolad commented on crates/full-node/sov-ethereum/src/handlers/subscribe/service.rs:
```diff
@@ -0,0 +1,173 @@
+            let mut state = self.ethereum.api_state_accessor();
+            let pending_block = self.evm.pending_block(&mut state);
+
+            for block_number in block_watermark.advance(pending_block.header.number - 1) {
```
Nit Let's use an exclusive range for Watermark. This will allow us to avoid subtracting 1 every time we call `Watermark::new` or `Watermark::advance` (currently happening four times across the two methods above).

LogvinovLeon replied:
Fixed

LogvinovLeon committed:
Refactor Watermark to use Range instead of RangeInclusive

- Change from RangeInclusive to Range for clearer semantics
- Rename 'mark' field to 'last_processed' for clarity
- Rename 'new_value' parameter to 'latest'
- Update EMPTY_RANGE to use 0..0 instead of 1..=0
- Update all tests and documentation accordingly

LogvinovLeon committed:
Fix clippy

LogvinovLeon committed:
Update Watermark to use RangeTo and improve documentation

- Change Watermark internal state from u64 to RangeTo<u64> for clearer semantics
- Update all tests to use RangeTo syntax (..n)
- Fix range calculations to properly handle exclusive ranges
- Improve documentation to reflect that watermark tracks and emits new items
- Update call sites in service.rs to use RangeTo syntax
- Remove outdated comment about monotonically increasing sequences

bkolad commented on crates/full-node/sov-ethereum/src/handlers/subscribe/service.rs:
```diff
@@ -0,0 +1,172 @@
+            let mut state = self.ethereum.api_state_accessor();
+            let pending_block = self.evm.pending_block(&mut state);
+
+            for tx_idx in tx_watermark.advance(..pending_block.transactions.end) {
```
This PR isn’t blocked on this, so feel free to merge it if you prefer this approach. I believe u64 would be perfectly fine here.

LogvinovLeon committed:
Fix tests
